File: python/40_CombinationSumII.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        all_sets = set()
        def makePermutations(cands, current_sum, current_set):
            if current_sum == target:
                current_set.sort()
                tuplified_current_set = tuple(current_set)
                if tuplified_current_set not in all_sets:
                    result.append(current_set)
                    all_sets.add(tuplified_current_set)
            elif current_sum > target:
                return result
            else:
                for i in range(len(cands)):
                    if cands[i] > target:
                        break
                    elif i == 0 or cands[i] != cands[i-1]:
                        makePermutations(cands[:i]+cands[i+1:], current_sum+cands[i], current_set + [cands[i]])
            return result
        candidates.sort()
        return makePermutations(candidates, 0, [])

File: swift/40_CombinationSumII.swift
class Solution {
    func combinationSum2(_ cands: [Int], _ target: Int) -> [[Int]] {
        var result: [[Int]] = [[Int]]()
        var candidates = cands.sorted(by: {
            a, b in 
            return a < b
        })

        func combine(_ curr: [Int], _ sum: Int, _ idx: Int = 0) {
            var size: Int = curr.count
            if sum == target {
                result.append(curr)
                return
            } else if sum > target {
                return
            } 
            
            for i in idx..<candidates.count {
                var n: Int = candidates[i]
                
                // to ensure we consider duplicates once, cands[i] != cands[i-1] if i has moved beyond its original starting point idx in the recursion stack's loop
                if size > 0 && n < curr[size-1] || i != idx && candidates[i] == candidates[i-1] { continue }
                combine(curr + [n], sum + n, i + 1)
            }
        }
        combine([], 0, 0)
        return result
    }
}
